planner/core/exhaust_physical_plans.go

```diff
@@ -1272,6 +1272,7 @@ func (ijHelper *indexJoinBuildHelper) removeUselessEqAndInFunc(idxCols []*expres
 				ijHelper.curPossibleUsedKeys = append(ijHelper.curPossibleUsedKeys, idxCols[idxColPos])
 				continue
 			}
+			ijHelper.curIdxOff2KeyOff[idxColPos] = -1
```

Comment:
Why we did this change? Or should we strengthen the case to `STRING` type only?

Reply:
if we didn't change this

for table `t1 (a int not null, b varcharbinary not null,index (a, b))` and `t2(a int not null, b varchar not null, index(a, b))`

the query: 't1 join t2 on t1. a = tb2.a and t2.a = t2.b'

will cause `ranges` only contain column `a`'s value(caused by https://github.com/pingcap/tidb/pull/21201), but `keyOff2IdxOff` include both `a` and `b`'s offset in

https://github.com/pingcap/tidb/blob/bfdad7c75210a30c796d82ccd882f381c2c20478/executor/builder.go#L3677

so the query will panic

@winoros is any other advise to fix this~? I'm not familiar with those part logic 😂